### DISCUSSION TITLE
vite4 lib how to package a vue sfc into one file

### DISCUSSION THREAD

**Discussion by zkisme** [vitejs/vite-plugin-vue#125] [Q&A]
### Related plugins

- [X] [plugin-vue](https://github.com/vitejs/vite-plugin-vue/tree/main/packages/plugin-vue)

- [ ] [plugin-vue-jsx](https://github.com/vitejs/vite-plugin-vue/tree/main/packages/plugin-vue-jsx)

### Describe the bug

I use this vite config 
```import { defineConfig } from 'vite';
import vue from '@vitejs/plugin-vue';

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [vue()],
  build: {
    lib: {
      formats: ['es'],
      entry: './src/index.js',
      name: 'xxx',
      fileName: '[name]',
    },
    rollupOptions: {
      external: ['vue'],
      output: {
        preserveModules: true,
      },
    },
  },
});
```

I want to output `HelloWorld.vue` to the `/components/helloworld.js` file.
But got the following files, It contains `HelloWorld.vue.js` and `HelloWorld.vue2.js`,
how to comple into one `HelloWorld.js` file

```
dist/es/components/HelloWorld.vue2.js
dist/es/_virtual/_plugin-vue_export-helper.js
dist/es/components/HelloWorld.vue.js
```

### Reproduction

https://stackblitz.com/edit/vitejs-vite-zktjzh?file=vite.config.js

### Steps to reproduce

npm run build

### System Info

```shell
System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Memory: 0 Bytes / 0 Bytes
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 16.14.2 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 7.17.0 - /usr/local/bin/npm
  npmPackages:
    @vitejs/plugin-vue: ^4.0.0 => 4.0.0 
    vite: ^4.1.4 => 4.1.4
```

### Used Package Manager

npm

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitejs/vite/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitejs/vite/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitejs.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitejs/vite-plugin-vue/issues) that reports the same bug to avoid creating a duplicate.
- [X] Make sure this is a Vite issue and not a framework-specific issue. For example, if it's a Vue SFC related bug, it should likely be reported to [vuejs/core](https://github.com/vuejs/core) instead.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitejs/vite-plugin-vue/discussions) or join our [Discord Chat Server](https://chat.vitejs.dev/).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by andrey-hohlov** [comment]
I stuck with the same issue building shared components library.

_Also I'm looking for a way to keep css file import in result file._

**Reply by agileago** [reply]
is there a solution ? @andrey-hohlov

**Comment by eric-gitta-moore** [comment]
For any friends who have come here. Here I would like to share my own solution. 

1. The issue with `_plugin-vue_export-helper.js` is not significant. It is almost impossible to delete (unless manually split into packages).
2. I really dislike the file `HelloWorld.vue2.js`, which causes the result of `vue-tsc` to not be able to correspond one-to-one with the `.js` output file. 

The main issue is to be addressed. Please try using `@vitejs/plugin-vue@^2`, and it should solve the problem. However, the reason for this issue remains to be investigated.

**Reply by eric-gitta-moore** [reply]
This line of code was directly caused by the pull request: https://github.com/vitejs/vite/pull/7909 

As of the current version, the modified code can be found here: https://github.com/vitejs/vite-plugin-vue/blob/dc1a3da0f3103975a8c225ff63c3a08e7e58b319/packages/plugin-vue/src/script.ts#L113-L130

```ts
if (lang === 'ts' && options.devServer) {
    return true
  }
```

We must enter the conditional statement `lang === 'ts' && options.devServer'`

One possible hacking method is

```ts
const noop = () => null;
export default defineConfig((env) => {
  return {
    plugins: [
      vue({ devServer: { watcher: { on: noop }, config: { server: { hmr: false } } } }),
      jsx(),
    ],
  };
});
```

Then we will be able to easily remove files like `xxx.vue2.js`
